### DISCUSSION TITLE
PowerShell one-step install script

### DISCUSSION THREAD

**Discussion by halr9000** [hlky/stable-diffusion#211] [Ideas]
I'd like to see a single PowerShell script to [do all the setup things](https://github.com/hlky/stable-diffusion/wiki/Installation#windows---step-by-step-installation-guide). Unsure if I can make the time to build it, I'll gauge interest here with what goals would look like and see what y'all think.

Initial goals [draft]:

- Handles first-time setup on Windows in a single step
- Best effort quick upgrade (maybe git refresh would just work)
- Installs all dependencies without any GUI steps. Includes git, python, conda, sd source, models
- Installs the optional face fix & upscale features by default (those are my favorite part of this project @hlky !)
- Switch from antiquated batch scripts to PowerShell (I am a bit biased, google "halr9000 powershell")
- Automate model downloads using [huggingface security tokens](https://huggingface.co/docs/hub/security-tokens). (Credit to [this sd gui colab](https://colab.research.google.com/drive/1NfgqublyT_MWtR5CsmrgmdnkWiijF3P3#scrollTo=dMjPceYlRG4C) for the idea.)

And version 2 might look like:

- Handles various upgrade scenarios and provide other options based on feedback

And alongside, something I've been wanting to do all year (going back to VQGAN days) is to make PowerShell really shine for AI art. I want to fork upstream and wrap/replace some of the SD python sample scripts with PowerShell versions, taking advantage of the features it has such as tab-completion and comprehensive help system. I like Python, but it's much easier to make an easy-to-use and learn, actually friendly CLI experience with PowerShell (that being why they made it).

What do y'all think?

**Comment by halr9000** [comment]
tagging @lex-drl @seanwcom for comments

**Comment by hlky** [comment]
Yes this is planned, if anyone wants to tackle it that would be appreciated.

> Automate model downloads using [huggingface security tokens](https://huggingface.co/docs/hub/security-tokens).

I appreciate you including this and not some unofficial links, to note: we don't want to support anything that bypasses the requirement of registering on huggingface and accepting the licenses for the model, I don't see why anyone can't just register on huggingface, accept the licenses and get a token to use.

**Comment by Lex-DRL** [comment]
@halr9000 not sure why I've been tagged, I barely contributed to this repo.

I support the idea, but it seems to me that it would be hard to detect all the possible combinations of whether conda/git is installed and if it has the right version. What if a user has a portable git and it's manually added to `PATH`? What if it's in a folder that only admin has write permissions to? What if this `git.exe` is there, but it has an outdated version?

As for single-step installation, it seems too limiting for power users. It should at least give an option to specify where to install conda, if it needs to be installed at all, etc.

Basically, if it's done right, this installation script will end up a mini package manager.

**Comment by JoshuaKimsey** [comment]
Very nice! It's basically a Windows version of what my [Linux Bash](https://github.com/JoshuaKimsey/Linux-StableDiffusion-Script) script does. I especially like the automatic model download. I've been considering how to implement that in an official way as well. Security token should be easy enough to deal with, my main concern was how slow it was to download the model initially. Perhaps it's just because HF was under a lot of strain that first day, but it took an hour to download that model, and I'm on fibre. I was worried doing that in the script could be too... bulky. But it may be that now it's normalized and won't take as long. Regardless, definitely a feature goal for an automatic installation to have, look forward to seeing your script! :)
